### PR TITLE
[FLOC-1281] Control service AMP server

### PR DESCRIPTION
**Pull request by itamarst**
https://clusterhq.atlassian.net/browse/FLOC-1281

### CODE REVIEW

File: flocker/control/_persistence.py [hunk 1]
```diff
@@ -62,6 +63,15 @@ def startService(self):
             self._deployment = Deployment(nodes=frozenset())
             self._sync_save(self._deployment)
 
+    def register(self, change_callback):
+        """
+        Register a function to be called whenever the configuration changes.
+
+        :param change_callback: Callabke that takes no arguments, will be
```
Comment (wallrj) [inline flocker/control/_persistence.py]: typo "Callabke"

File: flocker/control/_protocol.py [hunk 2]
```diff
@@ -107,10 +111,101 @@ def version(self):
...
+        """
+        An existing connection has been disconnected.
+
+        :param ControlAMP connection: The lost connection.
+        """
+        self.connections.remove(connection)
+
+    def node_changed(self, hostname, node_state):
+        """
+        We've received a node state updated from a connected client.
```
Comment (wallrj) [inline flocker/control/_protocol.py]: typo "updated"

File: flocker/control/test/test_protocol.py [hunk 3]
```diff
@@ -111,16 +115,76 @@ def test_deployment(self):
...
+        cluster_state = self.control_amp_service.cluster_state.as_deployment()
+        self.assertEqual(
+            sent[0],
+            (((ClusterStatusCommand,),
+              dict(configuration=TEST_DEPLOYMENT,
+                   state=cluster_state))))
+
+    def test_connection_lost(self):
+        """
+        When a connection is lost the ``ControlAMP`` is removed the services
```
Comment (wallrj) [inline flocker/control/test/test_protocol.py]: Typo "removed *from* the service's"

File: flocker/control/test/test_protocol.py [hunk 4]
```diff
@@ -111,16 +115,76 @@ def test_deployment(self):
...
+        set of connections.
+        """
+        marker = object()
+        self.control_amp_service.connections.add(marker)
+        self.patch(self.protocol, "callRemote", lambda *args, **kwargs: None)
```
Comment (wallrj) [inline flocker/control/test/test_protocol.py]: ``callRemote`` should return a Deferred. Does it matter that the patch function doesn't?

File: flocker/control/test/test_protocol.py [hunk 5]
```diff
@@ -111,16 +115,76 @@ def test_deployment(self):
...
+        self.protocol.makeConnection(StringTransport())
+        self.protocol.connectionLost(Failure(ConnectionLost()))
```
Comment (wallrj) [inline flocker/control/test/test_protocol.py]: The test doesn't check that this second protocol is ever added to the connections ``set``, but I suppose that's verified by the test above.

File: flocker/control/_protocol.py [hunk 6]
```diff
@@ -107,10 +111,101 @@ def version(self):
...
+    def startService(self):
+        self.endpoint_service.startService()
+
+    def stopService(self):
+        self.endpoint_service.stopService()
+        for connection in self.connections:
+            connection.transport.loseConnection()
+
+    def _send_state_to_connections(self, connections):
+        print "sending", connections
```
Comment (wallrj) [inline flocker/control/_protocol.py]: Remove debug statements.

File: flocker/control/_protocol.py [hunk 7]
```diff
@@ -107,10 +111,101 @@ def version(self):
...
+        :param endpoint: Endpoint to listen on.
+        """
+        self.connections = set()
+        self.cluster_state = cluster_state
+        self.configuration_service = configuration_service
+        self.endpoint_service = StreamServerEndpointService(
+            endpoint, ServerFactory.forProtocol(lambda: ControlAMP(self)))
+        # When configuration changes, notify all connected clients:
+        self.configuration_service.register(
+            lambda: self._send_state_to_connections(self.connections))
```
Comment (wallrj) [inline flocker/control/_protocol.py]: In FLOC-1281 you say that you still need to do part 3( "Notify all connected agents via AMP protocol whenever desired configuration changes.") but this looks like that part.

**Comment by wallrj** [comment]
Thanks @itamarst 

And sorry this took me so long to review. I think I get it...and it all looks fine to me.

* *all* the builds are green !!
* Coverage seems complete.

I left some inline comments. @exarkun made some comments about "IBoxReceiver".

Please address or answer those and merge.

File: flocker/control/_protocol.py (hunk 7 above)
Comment (itamarst) [inline flocker/control/_protocol.py]: Yeah, that was note to myself; I finished it up later before putting it up for review.

File: flocker/control/test/test_protocol.py (hunk 4 above)
Comment (itamarst) [inline flocker/control/test/test_protocol.py]: Hm. Yeah, probably will eventually, I'll fix that.

**Comment by wallrj** [comment]
I was a bit puzzled by what I got when I attempted to telnet to the agent-port. Here's a transcript:
```
(1281)(control-service-AMP-FLOC-1281 ? =)[~/.../HybridLogic/flocker]$ flocker-control --data-path=/tmp
sending [<ControlAMP at 0x2163cd0>]
Unhandled error in Deferred:
Unhandled Error
Traceback (most recent call last):
Failure: twisted.protocols.amp.TooLong: 
```

```
(control-service-AMP-FLOC-1281 ? =)[~/.../HybridLogic/flocker]$ telnet localhost 4524
hello
Trying ::1...
telnet: connect to address ::1: Connection refused
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
configuration�ccopy_regtatusCommand
_reconstructor
p0
(cflocker.control._model
Deployment
p1
c__builtin__
object
p2
Ntp3
Rp4
(dp5
S'nodes'
p6
c__builtin__
frozenset
p7
((lp8
tp9
Rp10
sb.state�ccopy_reg
_reconstructor
p0
(cflocker.control._model
Deployment
p1
c__builtin__
object
p2
Ntp3
Rp4
(dp5
S'nodes'
p6
c__builtin__
frozenset
p7
((lp8
tp9
Rp10
sb.Connection closed by foreign host.
```

I know it's pickle format, but didn't understand why it was being returned at all, given that I hadn't made a valid request.

**Comment by exarkun** [comment]
Use `nc` instead of `telnet`.  telnet is its own protocol and is not just a raw pipe between your keyboard and the remote host.  `nc` is a raw pipe.

**Comment by itamarst** [comment]
AMP isn't request/response, it's two way... and when you connect to control service it automatically sends you desired and actual state. Which currently involves pickle until FLOC-1240 is fixed IIRC. The "too long" thing is a problem.

**Pull request merged**